baybe/recommenders/pure/nonpredictive/clustering.py

```diff
@@ -1,12 +1,13 @@
 """Recommendation strategies based on clustering."""
 
 from abc import ABC
-from typing import ClassVar, List, Type, TypeVar
+from typing import ClassVar, List, Type, Union
 
 import numpy as np
 import pandas as pd
 from attrs import define, field
 from scipy.stats import multivariate_normal
+from sklearn.base import ClusterMixin
```

Comment:
What exactly is this?

Reply:
its one of the base classes of Sklearn, my hope was that it implements all functions that are used in thhis file so that mypyr ecognizes they exists

the reality unfirtunately is that we use a mixed set of cufunctions not part of any single sklearn base class

as a result the very proper thing here to do is create a `Protocol` although Adrian seems to have another idea as well, so Id suggest to discuss that when hes back, this code is functional